File: src/core/events.ts
import type {
  WorkspaceClaimResponse,
  WorkspaceClaimWindow,
  ClaimStatus,
  CoordinatorWorkflowDecision,
  TaskDispatch,
  WorkspaceActivity,
  WorkspaceMember,
  WorkspaceSpec,
  WorkspaceVisibility,
  WorkspaceWorkflowVoteResponse,
  WorkspaceWorkflowVoteWindow,
} from './types.js';

export interface BaseWorkspaceEvent {
  timestamp: string;
  workspaceId: string;
}

export interface WorkspaceStartedEvent extends BaseWorkspaceEvent {
  type: 'workspace.started';
  spec: WorkspaceSpec;
}

export interface WorkspaceInitializedEvent extends BaseWorkspaceEvent {
  type: 'workspace.initialized';
  sessionId?: string;
  availableAgents: string[];
  availableTools: string[];
  availableCommands?: string[];
}

export interface WorkspaceStateChangedEvent extends BaseWorkspaceEvent {
  type: 'workspace.state.changed';
  state: 'idle' | 'running' | 'requires_action' | 'closed';
}

export interface WorkspaceMessageEvent extends BaseWorkspaceEvent {
  type: 'message';
  role: 'user' | 'assistant' | 'system';
  text: string;
  visibility?: WorkspaceVisibility;
  memberId?: string;
  sessionId?: string;
  parentToolUseId?: string | null;
  raw: unknown;
}

export interface MemberRegisteredEvent extends BaseWorkspaceEvent {
  type: 'member.registered';
  member: WorkspaceMember;
}

export interface MemberStateChangedEvent extends BaseWorkspaceEvent {
  type: 'member.state.changed';
  member: WorkspaceMember;
}

export interface DispatchClaimedEvent extends BaseWorkspaceEvent {
  type: 'dispatch.claimed';
  dispatch: TaskDispatch;
  member: WorkspaceMember;
  claimStatus: ClaimStatus;
  note?: string;
}

export interface ClaimWindowOpenedEvent extends BaseWorkspaceEvent {
  type: 'claim.window.opened';
  claimWindow: WorkspaceClaimWindow;
}

export interface ClaimResponseEvent extends BaseWorkspaceEvent {
  type: 'claim.response';
  claimWindowId: string;
  response: WorkspaceClaimResponse;
}

export interface ClaimWindowClosedEvent extends BaseWorkspaceEvent {
  type: 'claim.window.closed';
  claimWindow: WorkspaceClaimWindow;
  responses: WorkspaceClaimResponse[];
  selectedRoleIds: string[];
}

export interface WorkflowVoteWindowOpenedEvent extends BaseWorkspaceEvent {
  type: 'workflow.vote.opened';
  coordinatorDecision: CoordinatorWorkflowDecision;
  voteWindow: WorkspaceWorkflowVoteWindow;
}

export interface WorkflowVoteResponseEvent extends BaseWorkspaceEvent {
  type: 'workflow.vote.response';
  voteId: string;
  response: WorkspaceWorkflowVoteResponse;
}

export interface WorkflowVoteWindowClosedEvent extends BaseWorkspaceEvent {
  type: 'workflow.vote.closed';
  coordinatorDecision: CoordinatorWorkflowDecision;
  voteWindow: WorkspaceWorkflowVoteWindow;
  responses: WorkspaceWorkflowVoteResponse[];
  approved: boolean;
}

export interface WorkflowStartedEvent extends BaseWorkspaceEvent {
  type: 'workflow.started';
  coordinatorDecision: CoordinatorWorkflowDecision;
  voteWindow?: WorkspaceWorkflowVoteWindow;
  nodeId?: string;
  stageId?: string;
}

export interface WorkflowStageEvent extends BaseWorkspaceEvent {
  type: 'workflow.stage.started' | 'workflow.stage.completed';
  nodeId: string;
  stageId?: string;
  roleId?: string;
}

export interface ActivityPublishedEvent extends BaseWorkspaceEvent {
  type: 'activity.published';
  activity: WorkspaceActivity;
}

export interface DispatchQueuedEvent extends BaseWorkspaceEvent {
  type: 'dispatch.queued';
  dispatch: TaskDispatch;
}

export interface DispatchStartedEvent extends BaseWorkspaceEvent {
  type: 'dispatch.started';
  dispatch: TaskDispatch;
  taskId: string;
  description: string;
}

export interface DispatchProgressEvent extends BaseWorkspaceEvent {
  type: 'dispatch.progress';
  dispatch: TaskDispatch;
  taskId: string;
  description: string;
  summary?: string;
  lastToolName?: string;
}

export interface DispatchCompletedEvent extends BaseWorkspaceEvent {
  type: 'dispatch.completed' | 'dispatch.failed' | 'dispatch.stopped';
  dispatch: TaskDispatch;
  taskId: string;
  outputFile: string;
  summary: string;
}

export interface DispatchResultEvent extends BaseWorkspaceEvent {
  type: 'dispatch.result';
  dispatch: TaskDispatch;
  taskId: string;
  resultText: string;
}

export interface ToolProgressEvent extends BaseWorkspaceEvent {
  type: 'tool.progress';
  taskId?: string;
  toolName: string;
  elapsedTimeSeconds: number;
}

export interface ResultEvent extends BaseWorkspaceEvent {
  type: 'result';
  subtype: string;
  result?: string;
  isError: boolean;
  sessionId: string;
  raw: unknown;
}

export interface ErrorEvent extends BaseWorkspaceEvent {
  type: 'error';
  error: Error;
}

export type WorkspaceEvent =
  | WorkspaceStartedEvent
  | WorkspaceInitializedEvent
  | WorkspaceStateChangedEvent
  | WorkspaceMessageEvent
  | MemberRegisteredEvent
  | MemberStateChangedEvent
  | DispatchClaimedEvent
  | ClaimWindowOpenedEvent
  | ClaimResponseEvent
  | ClaimWindowClosedEvent
  | WorkflowVoteWindowOpenedEvent
  | WorkflowVoteResponseEvent
  | WorkflowVoteWindowClosedEvent
  | WorkflowStartedEvent
  | WorkflowStageEvent
  | ActivityPublishedEvent
  | DispatchQueuedEvent
  | DispatchStartedEvent
  | DispatchProgressEvent
  | DispatchCompletedEvent
  | DispatchResultEvent
  | ToolProgressEvent
  | ResultEvent
  | ErrorEvent;

File: rust/crates/multi-agent-protocol/src/events.rs
use serde::{Deserialize, Serialize};

use crate::{
    ClaimStatus, CoordinatorWorkflowDecision, TaskDispatch, WorkspaceActivity,
    WorkspaceClaimResponse, WorkspaceClaimWindow, WorkspaceMember, WorkspaceSpec, WorkspaceStatus,
    WorkspaceVisibility, WorkspaceWorkflowVoteResponse, WorkspaceWorkflowVoteWindow,
};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct BaseWorkspaceEvent {
    pub timestamp: String,
    pub workspace_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum WorkspaceEvent {
    WorkspaceStarted {
        timestamp: String,
        workspace_id: String,
        spec: WorkspaceSpec,
    },
    WorkspaceInitialized {
        timestamp: String,
        workspace_id: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        session_id: Option<String>,
        available_agents: Vec<String>,
        available_tools: Vec<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        available_commands: Option<Vec<String>>,
    },
    WorkspaceStateChanged {
        timestamp: String,
        workspace_id: String,
        state: WorkspaceStatus,
    },
    MemberRegistered {
        timestamp: String,
        workspace_id: String,
        member: WorkspaceMember,
    },
    MemberStateChanged {
        timestamp: String,
        workspace_id: String,
        member: WorkspaceMember,
    },
    Message {
        timestamp: String,
        workspace_id: String,
        role: String,
        text: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        visibility: Option<WorkspaceVisibility>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        member_id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        session_id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        parent_tool_use_id: Option<String>,
    },
    DispatchQueued {
        timestamp: String,
        workspace_id: String,
        dispatch: TaskDispatch,
    },
    DispatchClaimed {
        timestamp: String,
        workspace_id: String,
        dispatch: TaskDispatch,
        member: WorkspaceMember,
        claim_status: ClaimStatus,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        note: Option<String>,
    },
    ClaimWindowOpened {
        timestamp: String,
        workspace_id: String,
        claim_window: WorkspaceClaimWindow,
    },
    ClaimResponse {
        timestamp: String,
        workspace_id: String,
        claim_window_id: String,
        response: WorkspaceClaimResponse,
    },
    ClaimWindowClosed {
        timestamp: String,
        workspace_id: String,
        claim_window: WorkspaceClaimWindow,
        responses: Vec<WorkspaceClaimResponse>,
        selected_role_ids: Vec<String>,
    },
    WorkflowVoteOpened {
        timestamp: String,
        workspace_id: String,
        coordinator_decision: CoordinatorWorkflowDecision,
        vote_window: WorkspaceWorkflowVoteWindow,
    },
    WorkflowVoteResponse {
        timestamp: String,
        workspace_id: String,
        vote_id: String,
        response: WorkspaceWorkflowVoteResponse,
    },
    WorkflowVoteClosed {
        timestamp: String,
        workspace_id: String,
        coordinator_decision: CoordinatorWorkflowDecision,
        vote_window: WorkspaceWorkflowVoteWindow,
        responses: Vec<WorkspaceWorkflowVoteResponse>,
        approved: bool,
    },
    WorkflowStarted {
        timestamp: String,
        workspace_id: String,
        coordinator_decision: CoordinatorWorkflowDecision,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        vote_window: Option<WorkspaceWorkflowVoteWindow>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        node_id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        stage_id: Option<String>,
    },
    WorkflowStageStarted {
        timestamp: String,
        workspace_id: String,
        node_id: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        stage_id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        role_id: Option<String>,
    },
    WorkflowStageCompleted {
        timestamp: String,
        workspace_id: String,
        node_id: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        stage_id: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        role_id: Option<String>,
    },
    DispatchStarted {
        timestamp: String,
        workspace_id: String,
        dispatch: TaskDispatch,
        task_id: String,
        description: String,
    },
    DispatchProgress {
        timestamp: String,
        workspace_id: String,
        dispatch: TaskDispatch,
        task_id: String,
        description: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        summary: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        last_tool_name: Option<String>,
    },
    DispatchCompleted {
        timestamp: String,
        workspace_id: String,
        dispatch: TaskDispatch,
        task_id: String,
        output_file: String,
        summary: String,
    },
    DispatchFailed {
        timestamp: String,
        workspace_id: String,
        dispatch: TaskDispatch,
        task_id: String,
        output_file: String,
        summary: String,
    },
    DispatchStopped {
        timestamp: String,
        workspace_id: String,
        dispatch: TaskDispatch,
        task_id: String,
        output_file: String,
        summary: String,
    },
    DispatchResult {
        timestamp: String,
        workspace_id: String,
        dispatch: TaskDispatch,
        task_id: String,
        result_text: String,
    },
    ActivityPublished {
        timestamp: String,
        workspace_id: String,
        activity: WorkspaceActivity,
    },
    ToolProgress {
        timestamp: String,
        workspace_id: String,
        tool_name: String,
        elapsed_time_seconds: u64,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        task_id: Option<String>,
    },
    Result {
        timestamp: String,
        workspace_id: String,
        subtype: String,
        is_error: bool,
        session_id: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        result: Option<String>,
    },
    Error {
        timestamp: String,
        workspace_id: String,
        error: String,
    },
}
